### PR TITLE
KEYCLOAK-18466 Configure HTTP client timeouts for adapters

### PR DESCRIPTION
**Pull request by mabartos**
JIRA: [KEYCLOAK-18466](https://issues.redhat.com/browse/KEYCLOAK-18466)

IMHO, it should be possible to customize timeouts for HTTP client included in the adapter core because each application/servlet requires various requirements. You can define these timeouts in XML (SAML) and JSON (OIDC) adapter configuration file. From QE perspective, some tests executed on particular machines require change of these timeouts in order to run properly.

@hmlnarik @mhajas Could you please take a look at it?

### CODE REVIEW

File: adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+                    You can create this truststore by extracting the public certificate of the IDP's SSL keystore.
+                </xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+        <xs:attribute name="truststorePassword" type="xs:string" use="optional">
+            <xs:annotation>
+                <xs:documentation>Password for the truststore keystore.</xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+        <xs:attribute name="socketTimeout" type="xs:long" default="-1">
```
Comment (mhajas) [inline adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd:517]: Do we need `long` here? 9223372036854775807 sounds like an overkill for the socket timeout. Maybe, we should follow the [API of httpclient](https://hc.apache.org/httpclient-legacy/apidocs/org/apache/commons/httpclient/params/HttpConnectionParams.html#setSoTimeout(int)) and use int?

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java [hunk 2]
```diff
@@ -376,6 +376,14 @@ public HttpClient build(AdapterHttpClientConfig adapterConfig) {
 
         configureProxyForAuthServerIfProvided(adapterConfig);
 
+        if (socketTimeout == -1 && adapterConfig.getSocketTimeout() != -1) {
```
Comment (mhajas) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java:379]:
```suggestion
        if (socketTimeout == -1 && adapterConfig.getSocketTimeout() > 0) {
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/DemoServletsAdapterTest.java [hunk 3]
```diff
@@ -210,99 +212,104 @@
     @Rule
     public AssertEvents assertEvents = new AssertEvents(this);
 
+    private static final Consumer<AdapterConfig> GLOBAL_ADAPTER_CONFIG = config -> {
```
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/DemoServletsAdapterTest.java:215]: Is there some instability in this test? I don't see it from the Jira? Should we link another Jira to the PR?

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @mabartos. Overall, the changes looks good to me, however I added few comments in the code. 

However, this is probably something we should discuss on mailing list, maybe system variable way of configuring this would be preferred as you suggested here: https://github.com/keycloak/keycloak/pull/8175.

File: adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+            </xs:element>
+            <xs:element name="PrincipalNameMapping" type="principal-name-mapping-type" minOccurs="0" maxOccurs="1">
+                <xs:annotation>
+                    <xs:documentation>When creating a Java Principal object that you obtain from methods like HttpServletRequest.getUserPrincipal(), you can define what name that is returned by the Principal.getName() method.</xs:documentation>
+                </xs:annotation>
+            </xs:element>
+            <xs:element name="RoleIdentifiers" type="role-identifiers-type" minOccurs="0" maxOccurs="1">
+                <xs:annotation>
+                    <xs:documentation>Defines what SAML attributes within the assertion received from the user should be used as role identifiers within the Java EE Security Context for the user.
+                        By default Role attribute values are converted to Java EE roles. Some IDPs send roles via a member or memberOf attribute assertion. You can define one or more Attribute elements to specify which SAML attributes must be converted into roles.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (mabartos) [reply adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd:59]: Ok.

File: adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd (hunk 1 above)
Comment (mabartos) [reply adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd:517]: I agree with you. I just wanted to keep the same type of the timeout variable and keep it consistent with our codebase. But it makes sense to use only integer type for this.

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java (hunk 2 above)
Comment (mabartos) [reply adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java:379]: Probably it's better condition. I assume, nobody wants to set socket timeout to 0; therefore in this case, the default value should be used.

File: adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd (hunk 1 above)
Comment (mabartos) [reply adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd:517]: I agree with you. I just wanted to keep the same type of the timeout variable and keep it consistent with our codebase. But it makes sense to use only integer type for this.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/DemoServletsAdapterTest.java (hunk 3 above)
Comment (mabartos) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/DemoServletsAdapterTest.java:215]: I've found out the instabilities, which this PR should solve are related to whole platform and there's created another issue (see #8175 ). Thus, this is out of date and I'll remove it.

**Comment by mabartos** [comment]
@mhajas Thanks for your comment. I'll fix this.

I've just started thread on Keycloak dev (See [here](https://groups.google.com/g/keycloak-dev/c/HLLo9wk0dSI)).

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java (hunk 2 above)
Comment (mhajas) [reply adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java:379]: It seems that zero is the default value which corresponds to infinite timeout.

File: adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd (hunk 1 above)
Comment (mhajas) [reply adapters/saml/core/src/main/resources/schema/keycloak_saml_adapter_1_13.xsd:517]: @mabartos I had a look at the long vs int again and I think we should revert to long, as we are using it in server code as well. The reason probably is, that it is also possible to configure TimeUnit which then decreases the maximum value. Also as you said, we should stay consistent. Sorry for the confusion.

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java [hunk 5]
```diff
@@ -113,9 +113,9 @@ public void checkServerTrusted(X509Certificate[] chain,
     protected long connectionTTL = -1;
     protected TimeUnit connectionTTLUnit = TimeUnit.MILLISECONDS;
     protected HostnameVerifier verifier = null;
-    protected long socketTimeout = -1;
+    protected int socketTimeout = -1;
```
Comment (hmlnarik) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java:116]: Why this change?

File: adapters/saml/core/src/main/java/org/keycloak/adapters/cloned/HttpClientBuilder.java [hunk 6]
```diff
@@ -112,9 +112,9 @@ public void checkServerTrusted(X509Certificate[] chain,
     protected long connectionTTL = -1;
     protected TimeUnit connectionTTLUnit = TimeUnit.MILLISECONDS;
     protected HostnameVerifier verifier = null;
-    protected long socketTimeout = -1;
+    protected int socketTimeout = -1;
```
Comment (hmlnarik) [inline adapters/saml/core/src/main/java/org/keycloak/adapters/cloned/HttpClientBuilder.java:115]: Same here - why this change?

File: adapters/saml/core/src/test/java/org/keycloak/adapters/saml/config/parsers/KeycloakSamlAdapterXMLParserTest.java [hunk 7]
```diff
@@ -117,57 +117,59 @@ public void testParseSimpleFileNoNamespace() throws Exception {
...
-        assertFalse(sp.isKeepDOMAssertion());
-        assertEquals(2, sp.getKeys().size());
+        assertThat(sp.getEntityID(), is("sp"));
+        assertThat(sp.getSslPolicy(), is("EXTERNAL"));
+        assertThat(sp.getNameIDPolicyFormat(), is("format"));
+        assertThat(sp.isForceAuthentication(), is(true));
+        assertThat(sp.isIsPassive(), is(true));
+        assertThat(sp.isAutodetectBearerOnly(), is(false));
+        assertThat(sp.isKeepDOMAssertion(), is(false));
+        assertThat(sp.getKeys().size(), is(2));
```
Comment (hmlnarik) [inline adapters/saml/core/src/test/java/org/keycloak/adapters/saml/config/parsers/KeycloakSamlAdapterXMLParserTest.java:131]: Nit:
```suggestion
        assertThat(sp.getKeys(), hasSize(2));
```

File: services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java [hunk 8]
```diff
@@ -134,8 +134,8 @@ private void lazyInit(KeycloakSession session) {
         if (httpClient == null) {
             synchronized(this) {
                 if (httpClient == null) {
-                    long socketTimeout = config.getLong("socket-timeout-millis", 5000L);
-                    long establishConnectionTimeout = config.getLong("establish-connection-timeout-millis", -1L);
+                    int socketTimeout = config.getInt("socket-timeout-millis", 5000);
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:137]: Why this change?

File: services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java [hunk 9]
```diff
@@ -101,9 +101,9 @@ public void checkServerTrusted(X509Certificate[] chain,
     protected long maxConnectionIdleTime = 900000;
     protected TimeUnit maxConnectionIdleTimeUnit = TimeUnit.MILLISECONDS;
     protected HostnameVerifier verifier = null;
-    protected long socketTimeout = -1;
+    protected int socketTimeout = -1;
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java:104]: Why this change?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java [hunk 10]
```diff
@@ -229,4 +247,55 @@ protected void setAdapterServletTimeOffset(int timeOffset, String servletUri) {
...
+            }
+        }
+        return file;
+    }
+
+    /**
+     * Custom Object mapper
+     */
+    private static ObjectMapper getSystemPropertiesObjectMapper() {
+        ObjectMapper mapper = new ObjectMapper(new SystemPropertiesJsonParserFactory());
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java:296]: `ObjectMapper` should be a static constant rather than recreated each time.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java [hunk 11]
```diff
@@ -65,15 +73,25 @@ protected static WebArchive servletDeploymentMultiTenant(String name, Class... s
     }
 
     protected static WebArchive servletDeployment(String name, Class... servletClasses) {
-        return servletDeployment(name, "keycloak.json", servletClasses);
+        return servletDeployment(name, DEFAULT_ADAPTER_CONFIG, servletClasses);
     }
 
-    protected static WebArchive servletDeployment(String name, String adapterConfig, Class... servletClasses) {
+    protected static WebArchive servletDeployment(String name, Consumer<AdapterConfig> adapterConfig, Class... servletClasses) {
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java:79]: It seems that `adapterConfig` here is always `null`. Where is this method used from? Why is the `Consumer` parameter introduced?

**Review by hmlnarik — COMMENTED** [review]
Thank you @mabartos. Could you please have a look at the few questions inline? It seems that some changes (changing type) are unrelated to the JIRA, is there any reason for doing so?
Also note that any new saml adapter xsd has to be published to the web, see https://github.com/keycloak/keycloak-web/tree/master/static/schema

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java (hunk 5 above)
Comment (mabartos) [reply adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java:116]: I've been discussing it with @mhajas and considering to change from _long_ to _int_ in order to follow Apache HTTP client API. However, as you've mentioned, this is unrelated to the JIRA and the change is not necessary though.

File: adapters/saml/core/src/main/java/org/keycloak/adapters/cloned/HttpClientBuilder.java (hunk 6 above)
Comment (mabartos) [reply adapters/saml/core/src/main/java/org/keycloak/adapters/cloned/HttpClientBuilder.java:115]: Same as above.

File: adapters/saml/core/src/test/java/org/keycloak/adapters/saml/config/parsers/KeycloakSamlAdapterXMLParserTest.java (hunk 7 above)
Comment (mabartos) [reply adapters/saml/core/src/test/java/org/keycloak/adapters/saml/config/parsers/KeycloakSamlAdapterXMLParserTest.java:131]: +1

File: services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java (hunk 8 above)
Comment (mabartos) [reply services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:137]: Same as [this](https://github.com/keycloak/keycloak/pull/8165#discussion_r658516576)

File: services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java (hunk 9 above)
Comment (mabartos) [reply services/src/main/java/org/keycloak/connections/httpclient/HttpClientBuilder.java:104]: Same as [this](https://github.com/keycloak/keycloak/pull/8165#discussion_r658516576).

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java (hunk 10 above)
Comment (mabartos) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java:296]: You're right.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java (hunk 11 above)
Comment (mabartos) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java:79]: In the first commit of this PR, this is used in _DemoServletAdapterTest_, but the solution for instability of the test is out of scope of this PR. So, it's like a preparation for another related tasks and configure adapter directly from a test. Please, see [here](https://github.com/keycloak/keycloak/pull/8165/commits/d1b77b938f54ef1d1070889eb020a79fdf21b503#diff-a7d3a845d41009122b21855f7fa779bbf324c4e6ba58f4404bf4f58addc89cadR215). @hmlnarik WDYT?

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java (hunk 5 above)
Comment (hmlnarik) [reply adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/HttpClientBuilder.java:116]: Please keep this PR minimal, it is already big enough.

**Comment by mabartos** [comment]
Created PR to _keycloak-web_. See https://github.com/keycloak/keycloak-web/pull/172

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java (hunk 11 above)
Comment (hmlnarik) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java:80]: The latter would be better, thank you

Comment (mabartos) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/AbstractServletsAdapterTest.java:80]: Ok

**Comment by mabartos** [comment]
@mhajas @hmlnarik Rebased and fixed all your suggestions. After review, these commits can be squashed. I'll execute tests in pipeline and I'll share the results.

**Comment by hmlnarik** [comment]
Does it make sense to also include the changes requested in #7968?

**Comment by mabartos** [comment]
@hmlnarik I've been thinking about it and I agree with it. I'll merge it.

**Comment by hmlnarik** [comment]
Thanks, @mabartos.

I only realized now that there also needs to be a documentation for the added properties:
- https://github.com/keycloak/keycloak-documentation/blob/master/securing_apps/topics/saml/java/general-config/idp_httpclient_subelement.adoc
- https://github.com/keycloak/keycloak-documentation/blob/master/securing_apps/topics/oidc/java/java-adapter-config.adoc
Apologies I missed this earlier.

**Comment by mabartos** [comment]
@hmlnarik Thanks. 
I merged the PR #7968. Thanks @ruromero for your contribution.

PR to _documentation_: https://github.com/keycloak/keycloak-documentation/pull/1220
PR to _keycloak-web_: https://github.com/keycloak/keycloak-web/pull/172

I'll execute the pipeline tests again. Previously, these tests passed in pipeline, therefore this PR didn't break anything.

**Comment by mabartos** [comment]
@hmlnarik The pipeline tests passed. See https://master-jenkins.com/job/universal-test-pipeline-adapters/506/

**Review by mhajas — CHANGES_REQUESTED** [review]
Thanks for all the changes @mabartos. I added one nitpick comment, however, I believe it makes sense to address it. WDYT?

File: adapters/oidc/adapter-core/src/test/java/org/keycloak/adapters/KeycloakDeploymentBuilderTest.java [hunk 12]
```diff
@@ -101,5 +105,18 @@ public void loadSecretJwtCredentials() {
...
+        assertThat(deployment, CoreMatchers.notNullValue());
+
+        HttpClient client = deployment.getClient();
+        assertThat(client, CoreMatchers.notNullValue());
 
+        long socketTimeout = client.getParams().getIntParameter(CoreConnectionPNames.SO_TIMEOUT, -2);
+        long connectionTimeout = client.getParams().getIntParameter(CoreConnectionPNames.CONNECTION_TIMEOUT, -2);
+
+        assertThat(socketTimeout, CoreMatchers.is(2000L));
+        assertThat(connectionTimeout, CoreMatchers.is(6000L));
```
[the remark this answers is not in the document]
Comment (mabartos) [reply adapters/oidc/adapter-core/src/test/java/org/keycloak/adapters/KeycloakDeploymentBuilderTest.java:120]: @mhajas Yeah, sure. Thanks

[the remark this answers is not in the document]
Comment (mabartos) [reply adapters/oidc/adapter-core/src/test/java/org/keycloak/adapters/KeycloakDeploymentBuilderTest.java:120]: @mhajas However, the `connectionTTL` is only used for internal purposes, especially to create _HttpClient_ with proper _Client Connection Manager_ and with connection pool size greater than 1. The `connectionTTL` attribute is not included in the HttpParams provided by Apache _http-core_ and it seems it shouldn't be revealed in the same manner like those timeouts. In this case, I don't want to create a new parameter which is out of scope of _http-core_ interface. WDYT?

[the remark this answers is not in the document]
Comment (mhajas) [reply adapters/oidc/adapter-core/src/test/java/org/keycloak/adapters/KeycloakDeploymentBuilderTest.java:120]: @mabartos I am not sure I am following. Could you elaborate a little bit more, please? What is the difference when the configuration option is used in `HttpParams` and in `ClientConnectionManager`? The `ThreadSafeClientConnManager` class comes from `httpclient` as well, so it is not our internal class. Also, we are using the `connectionTTL` option also for [HttpClientProvider](https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java#L141) of the Keycloak server.

Are you suggesting removing the configuration of connectionTTL from this PR completely?

[the remark this answers is not in the document]
Comment (mabartos) [reply adapters/oidc/adapter-core/src/test/java/org/keycloak/adapters/KeycloakDeploymentBuilderTest.java:120]: @mhajas In the `ThreadSafeClientConnManager` is the field `connectionTTL` with private access modifier and it's not possible to obtain it from the class and where it's only used for internal instantiating of pool entry. The field is completely hidden. Anyway, I'd like to find already existing parameter for `connectionTTL` in order to include it to the `HttpParams`, but I think, we will have to create a new one. 

However as you've suggested, if the support for the `connectionTTL` is present in this PR, there should be verified the proper settings of the property. I'll add a new parameter to `HttpParams` in order to obtain configuration externally. 
Thanks.

**Comment by mabartos** [comment]
@mhajas I've added _connectionTTL_ parameter to the parameters of the HTTP client. Pipeline tests passed (no new error appeared). See https://master-jenkins.com/job/universal-test-pipeline-adapters/519/

**Comment by mabartos** [comment]
After discussion with @mhajas we've decided to get rid off the custom variables for `HttpParams` and in a follow-up task to update whole test class `HttpClientBuilder` for adapters and provide particular test cases. Removed last commit and rebased with master.

**Comment by mabartos** [comment]
@mhajas JFYI the pipeline tests passed. See https://master-jenkins.redhat.com/job/universal-test-pipeline-adapters/534/

**Review by mhajas — DISMISSED** [review]
LGTM @mabartos, thanks for the changes.

**Comment by mabartos** [comment]
@hmlnarik JFYI: along with merging this PR, these PRs should be merged as well:

PR to documentation: [keycloak/keycloak-documentation#1220](https://github.com/keycloak/keycloak-documentation/pull/1220)
PR to keycloak-web: [keycloak/keycloak-web#172](https://github.com/keycloak/keycloak-web/pull/172)

**Comment by hmlnarik** [comment]
Thank @mabartos, great job! The changes look good to me apart from one inconsistency in letter case - there is already `proxyUrl` and the same letter case should be followed in the `connectionTtl` property (applies to all three PRs)

**Comment by sguilhen** [comment]
@mabartos Good work on this one. I just missed the new subsystem XSD for OIDC as7-eap6. Versions were bumped in the subsystem files but version 1_2 of the XSD is missing in this PR.

**Comment by mabartos** [comment]
> @mabartos Good work on this one. I just missed the new subsystem XSD for OIDC as7-eap6. Versions were bumped in the subsystem extension (https://github.com/keycloak/keycloak/pull/8165/files#diff-d9314783e37a970e7660f8ed687e723564fcaaefe4a9b5dc2d85607d1c7b43daR40) but version 1_2 of the XSD is missing in this PR.

@sguilhen Nice catch! Thanks.

**Comment by mabartos** [comment]
@hmlnarik @sguilhen I addressed your comments and rebased with the most recent master. 

After discussion with @hmlnarik, I changed the property names with suffix `-millis` for OIDC. 
For SAML, where the configuration of the HTTP client is defined in XML file, I didn't add the suffix (in this case the property `socketTimeout` would change to `socketTimeoutMillis`), because in the description of the fields, there are already defined the time units; milliseconds. Therefore, IMHO, it's quite counterproductive to define these units again.

I executed the pipeline tests, so the result will be lately.

**Comment by sguilhen** [comment]
@mabartos This renaming of the OIDC properties to include the **-millis** suffix creates a mismatch between the SAML properties that are named differently and the JSON properties in AdapterConfig

This means that when the SAML adapter subsystem reads the model and the adapter later on converts the subsystem config into JSON and tries to load AdapterConfig from the converted JSON we have an exception being thrown because the properties don't match. See https://issues.redhat.com/browse/KEYCLOAK-18302 for an example of an issue that I've recently fixed in which the SAML property and the JSON attribute in AdapterConfig had different names.

In the PR for that Jira you can see that I've added a test that configures the problematic property and then tries to convert it into JSON and load the JSON config into AdapterConfig. Without the correct mapping done here https://github.com/keycloak/keycloak/pull/8244/files#diff-a350e9551196cf925e4145d0a0c4e091f0ca2b3f31d28458ec248af121c44e1cR154 the test fails.

**Comment by mabartos** [comment]
@sguilhen Thanks for your comment. However, I'm not sure I understand. I'm not aware that SAML adapter converts adapter subsystem configuration to JSON. I'd say it's valid for OIDC adapters. As I saw in your attached PR, everything is related to OIDC, not SAML. For SAML adapter, HTTP client attributes are parsed in class _HttpClientParser_, which are then used in _HttpClientBuilder_ for building the client.

For OIDC Wildfly and AS7 subsystem, there are defined XML properties with the `-millis` suffix same as for JSON properties in _AdapterConfig_, so there shouldn't be any incompatibilities.

@sguilhen Could you please explain it more?

**Comment by sguilhen** [comment]
> @sguilhen Thanks for your comment. However, I'm not sure I understand. I'm not aware that SAML adapter converts adapter subsystem configuration to JSON. I'd say it's valid for OIDC adapters. As I saw in your attached PR, everything is related to OIDC, not SAML. For SAML adapter, HTTP client attributes are parsed in class _HttpClientParser_, which are then used in _HttpClientBuilder_ for building the client.
> 
> For OIDC Wildfly and AS7 subsystem, there are defined XML properties with the `-millis` suffix same as for JSON properties in _AdapterConfig_, so there shouldn't be any incompatibilities.
> 
> @sguilhen Could you please explain it more?
> 
> EDIT:
> I added test case for the WF subsystem to verify the JSON parsing.
> See [here](https://github.com/keycloak/keycloak/pull/8165/commits/685fc210d679e2f3332f83fb69bc40c46a7246d3#diff-45be67b9727261060237cf82a49a35364ab157920850906636d6be11f5943fcbR154).

Hi @mabartos You are right, I've got it the other way around, its the OIDC subsystem that converts the model to JSON, SAML works based on XML all the way. This looks good to me!

**Comment by mabartos** [comment]
@hmlnarik @mhajas @sguilhen I executed the pipeline tests again in order to ensure that everything works as expected. The pipeline tests passed. See https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/546/

**Review by hmlnarik — APPROVED** [review]
Approving based on @mhajas and @sguilhen 's reviews
